### ISSUE TITLE
[feature request] add* functions for leaflet package

### ISSUE THREAD

**Issue by tim-salabim** [leonawicz/tiler#4]
This is to continue the twitter discussion as per https://twitter.com/TimSalabim3/status/1006414006201077761

As you correctly point out in your response to my tweet, our (mapview's) primary interest is in seamless provision of tiled images for rendering on a leaflet map. The crux here is that the map should not be created as a side of the tile generation but that the generated tiles can be added to an existing leaflet object. Hence, the `add*` functionality. As I mention in my tweet we have just recently implemented such `add*` functions and along with those had to also implement functionality very similar to what **tiler** provides, i.e. tile generation. In fact, we simply had to replace [one line](https://github.com/r-spatial/mapview/blob/develop/R/extensions.R#L1236) in the respective functions to reproduce our functionality with that of tiler. This is great, as it means we can drop the heavy python dependency and replace it with a light(er)-weight R dependency. Taking this a little further, I think it would be beneficial for everyone to have those `add*` functions provided by tiler, rather than mapview. Specifically, I think that a) these functions address a rather special case which is somewhat on the brink of mapview's scope and b) having control over the whole process from tile creation to rendering is desirable in many ways. 

This, of course, introduces some questions to be considered for tiler:
First and foremost, it would mean to add a **htmlwidgets** and **leaflet** dependency/import. Secondly, it means that some javascript libs along with proper bindings would need to be added to get a smooth behaviour. Most of these are already in mapview and could easily be ported and adapted to work with tiler. I would be more than happy to help with this. 
On the other hand, this would open up possibilities to advance tiler functionality towards a fully functional renderer of non-geographic tiles as tiler could also provide functions to render (circle)markers/polylines/polygons on top of tiled images (via `leaflet-rastercoords` js binding). Again, I would be happy to help with those.

I know that this is a lot to digest, so I'm gonna stop here. In case anything is unclear, I am happy to elaborate. I do have some further suggestions/questions but I'm gonna reserve those for a later date as I think you will have enough to take in and consider for now.

**Comment by leonawicz** [comment]
I appreciate this info, thank you. It sounds like you are proposing ideas for adding some functionality to tiler while removing it from mapview? Please clarify if I'm off the mark. I looked at that line of code and it does look like you are considering leaving that work to tiler so you don't need to retain the python/python-gdal requirement for the many users of mapview who may have little to no interest in tile generation, is that right? (But whom could elect to jump through those hoops with respect to tiler if they actually want tiles)

I suppose an option is to add functions to tiler. Would this also mean that mapview would place tiler under Suggests in DESCRIPTION and perhaps anything that might call it would use requireNamespace to check for tiler, so that maview does not have to actually import it against the needs of most users? Perhaps none of that is an issue if the add* functions are in tiler.

I'm open to the idea, though not excited to add more libraries and dependencies, particularly if the only outcome is to add the ability to make maps with Leaflet, which I'd prefer to just have people do directly using the `leaflet` package itself. Even the auto-generated `preview.html` with Leaflet widget in tiler just relies on the minimum of JS and HTML hardcoded into an R function in order to avoid requiring the `leaflet` package for "presentation purposes", which I guess is the blanket term I'd use for "things on the edge of tiler's scope".

So it sounds like we have something here that is a bit on the periphery of scope for both our packages. mapview seems huge compared to tiler, and is clearly much more robust and time-tested and will surely remain more highly utilized. Do you think there is a way to move some of this functionality you are talking about to tiler, while still making "map-making stuff" optionally supported in tiler? I.e., is there any wiggle room for not making these absolute dependencies since they aren't needed for tiling, but only for tangential subsequent applications? For example, tiler currently does not import `jpeg` or `bmp`, so if someone tries to use a jpg or bmp file as an input instead of a png and they don't already have those packages installed, requireNamespace will return a message about the needed package. I didn't want to force include them since they are unlikely to be used much and don't contribute anything uniquely necessary to the tiling process.

I'll have to take a moment to play around and familiarize myself with these new functions in mapview. I'm a bit unclear on the part about adding markers etc. over non-geographic base map tiles. I've only used markers so far when using previously generated and hosted tile sets via the `leaflet` package, but they worked fine when just calling a `leaflet` function to add markers. But maybe we are not doing things in the same way. I have only been making tiles for base maps, not raster data overlays.

Anyway, let me know what you think and we can iterate on this and zero in on something together. I appreciate holding off on the other suggestions for now, I'm swamped with things in general.

Also, if you haven't seen it, check out  #2 regarding non-Mercator tile support. I think this is promising and would like this to be in the next version. The big breaking change would simply be deriving output tile CRS from input file CRS by default and not forcing reprojections to EPSG 4326.

Another option I've been thinking about that should be easy but just haven't gotten to, that I think people would really appreciate, is allowing the `file` argument to be an existing raster layer in the workspace. I suppose the argument name should be something like `x` instead of `file` if that becomes the case.

I appreciate your thoughts and assistance. :)
Matt

**Comment by leonawicz** [comment]
I'm still thinking it's a great idea for `tiler` to be the package that handles all things possible regarding non-geographic map tiles. I'd love that to become a core part of its scope. Mostly I'm unclear on what the proposed changes would mean in terms of the full weight of new dependencies. For example, addling a couple R packages is not as big a deal as adding more system requirements, but I'm not clear on all what would be involved at this point or a full understanding of the uses cases that will be served.

I did add a few simple convenience functions to the `projections` branch the other day, e.g. `tile_meta`. These pull basic metadata out of the `tilemapresource.xml` file, making it easier to pass the info to `leafletCRS`.

**Comment by tim-salabim** [comment]
If I am not mistaken, htmlwidgets would need to be at least in suggests and leaflet could be in enhances. I guess the biggest change would be the javascript libs and bindings. Have a look [here](https://github.com/r-spatial/mapview/tree/develop/inst/htmlwidgets/lib/leaflet-rastercoords) to get an idea about the extent. This is only for non-projected maps using `L-CRS.Simple`. For supporting projected tiles there may be some additional bindings needed, though I am not sure at this stage if any and how huge those would be.
I will not have too much time in the coming two weeks, but maybe I will find some time to provide a PR so you could get an idea about the extent of changes.

**Comment by leonawicz** [comment]
Oh, that doesn't look like much at all. I've already got (most of) that retained in there, used for my tiler previewer. See [here](https://github.com/leonawicz/tiler/blob/master/R/viewer.R#L84) and below.

Perhaps with other functions added, I might break this back out in some other form, but it's there now. I was imagining you meant gigantic heaps of source code and files far bigger and more numerous than the package itself. We can take it a step at a time too. I don't know what we'd need for other projections either. We're currently having difficulties on another issue thread, not with making tiles or even displaying them in Leaflet with the previewer, but simply getting them to show up in Leaflet via the `leaflet` package.

No worries on time. I also don't have much time for this around my other obligations. I don't have a goal set for when I even want to merge the `projections` branch into master and publish a new update. For the moment it will remain EPSG 4326 plus non-geographic on CRAN.

**Comment by tim-salabim** [comment]
What exactly are the problems you have with leaflet generation?

**Comment by leonawicz** [comment]
Below is a shorter example based on the example [here](https://github.com/jmlondon/naturalearth_3571) referenced in #2 

I put the tiles online, so you don't need to generate them. If you install the latest version of the `projections` branch, `tile_meta` now works with a URL string, so you can still read `tilemapresource.xml` in the tiles directory when the tiles are online.

The [preview.html](https://leonawicz.github.io/tempdocs/ne3571/preview.html) works fine. But we haven't gotten these tiles working with `leaflet` below.

```
devtools::install_github("leonawicz/tiler", ref = "projections")
library(tiler)
library(leaflet)

tiles <- "https://leonawicz.github.io/tempdocs/ne3571/tiles"
tile_url <- file.path(tiles, "{z}/{x}/{y}.png")
meta <- tile_meta(tiles)
b <- meta$bounds
center <- c(mean(c(b[[1]][2], b[[2]][2])), mean(c(b[[1]][1], b[[2]][1])))
            
crs <- leafletCRS(crsClass = "L.Proj.CRS", code = "EPSG:3571",
                  proj4def = "+proj=laea +lat_0=90 +lon_0=180 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs",
                  resolutions = meta$res, origin = meta$origin, bounds = b)

leaflet(options = leafletOptions(crs = crs)) %>% 
  addMarkers(lng = center[2], lat = center[1], label = "CENTER") %>%
  addTiles(urlTemplate = tile_url)
```

Looking at the structure of the leaflet object, it looks like all the information is correct. The map marker is right in the center as expected. Seems like the only issue is the tiles not displaying.

**Comment by leonawicz** [comment]
It also appears that if the tiles are added first, even though they do not display, this prevents the map marker from appearing when added second.

**Comment by tim-salabim** [comment]
Have you tried opening the map in the browser? Via the 'Open in broser' button above the viewer window? RStudio currently has issues with `https` protocols.

**Comment by leonawicz** [comment]
Yeah, no luck in the browser, or even outside RStudio entirely just using the standard R gui.

**Comment by tim-salabim** [comment]
Ok, I'll take a look at it tonite.

**Comment by tim-salabim** [comment]
Having a quick look, this is the best I can come up with so far:

```r
library(tiler)
library(leaflet)

tiles <- "https://leonawicz.github.io/tempdocs/ne3571/tiles"
tile_url <- file.path(tiles, "{z}/{x}/{y}.png")
meta <- tile_meta(tiles)
b <- meta$bounds
center <- c(mean(c(b[[1]][2], b[[2]][2])), mean(c(b[[1]][1], b[[2]][1])))

crs <- leafletCRS(crsClass = "L.Proj.CRS", code = "EPSG:3571",
                  proj4def = "+proj=laea +lat_0=90 +lon_0=180 +x_0=0 +y_0=0 +ellps=WGS84 +datum=WGS84 +units=m +no_defs",
                  resolutions = rev(c(286 * c(1, 2, 2^seq(2, 8)))),
                  bounds = b)

leaflet(options = leafletOptions(crs = crs)) %>% 
  addMarkers(lng = center[2], lat = center[1], label = "CENTER") %>%
  setView(lng = center[2], lat = center[1], zoom = 3) %>%
  addTiles(urlTemplate = tile_url, options = tileOptions(tileSize = 286,
                                                         minZoom = 0,
                                                         maxZoom = 9,
                                                         zoomOffset = 0,
                                                         maxNativeZoom = 8,
                                                         minNativeZoom = 3)) %>%
  mapview::addMouseCoordinates() # to get some info on mouse
```
You will see in the mouse coordinates strip at the top of the map that the region that is valid does not quite correspond to the region of the tiles. I think we need to include the `leaflet-rastercoords` lib and provide a proper binding so that we get the same result as in your preview

**Comment by tim-salabim** [comment]
This:
```r
library(tiler)
library(leaflet)

tiles <- "https://leonawicz.github.io/tempdocs/ne3571/tiles"
tile_url <- file.path(tiles, "{z}/{x}/{y}.png")
meta <- tile_meta(tiles)
b <- meta$bounds
center <- c(mean(c(b[[1]][2], b[[2]][2])), mean(c(b[[1]][1], b[[2]][1])))

crs = leafletCRS(
  crsClass = "L.CRS.Simple"
)

leaflet(options = leafletOptions(crs = crs)) %>% 
  addMarkers(lng = center[2], lat = center[1], label = "CENTER") %>%
  setView(lng = center[2], lat = center[1], zoom = 3) %>%
  addTiles(urlTemplate = tile_url, options = tileOptions(tileSize = 286,
                                                         minZoom = 0,
                                                         maxZoom = 14,
                                                         zoomOffset = 0,
                                                         maxNativeZoom = 8,
                                                         minNativeZoom = 3)) %>%
  mapview::addMouseCoordinates() # to get some info on mouse
```
is getting closer. If we can now have a set of proper leaflet-rastercoords bindings, then the marker should also be placed at the correct position.

**Comment by leonawicz** [comment]
I'm having similar success now, thanks. Particularly I've had some success with `crsClass = "L.Proj.CRS"`. My issue previously might have been related to improper `tileOptions`. Still some issues with zoom levels and points do not plot properly but I can see the map.

I don't know enough about this, but given the official support for projections in `leaflet`, my gut feeling is that trying to make a spatial projection like this one work via `crsClass = "L.CRS.Simple"` might not the "right" solution. It seems like there should still be a relatively straight forward way that avoids treating all non-4326 projections as simple CRS/non-spatial. Thoughts?

**Comment by tim-salabim** [comment]
Sorry for the long silence. I have now ported all functionality to add tiled imagery to a leaflet object as it was implemented in mapview to tiler. You can see and try in my fork. https://github.com/tim-salabim/tiler/blob/master/R/leaflet.R I had to copy over a few internal functions from mapview as well, along with Imports needed for those (e.g. lattice).
Maybe it is  a good idea to work on this fork until we feel that things are mature enough to be merged into tiler repo. 
Please let me know your thoughts on this.

**Comment by leonawicz** [comment]
No problem. I'm busy too. I just installed the forked version and tried:

```
library(tiler)
library(leaflet)
library(raster)
rst = raster(nrows = 5000, ncols = 10000, xmn = 0, xmx = 10000, ymn = -5000, ymx = 0)
rst[] = runif(ncell(rst))
map = leaflet(options = leaflet::leafletOptions(crs = leafletCRS(crsClass = "L.CRS.Simple")))
map %>% addTiledRasterImage(rst)
```

In RStudio no map appears (in browser view), just the gray background with the zoom buttons in the corner. In the standard R GUI, simply nothing happens after `addTiledRasterImage`.

Unfortunately it takes more than a minute just to generate the tiles. Nothing surprising there, but it's still hard for me to envision many use cases where either (1) there is still reason to tile something that is small enough that it processes quickly or (2) there is something that is large enough to necessitate tiling but is also worth waiting for in real time (like in a Shiny app). I could see an edge case where it's just part of offline interactive data exploration, though I do not think there would be a demand for that when a user in that case might be more likely to simply plot the raster layer.

It definitely a useful function (I'm just not able to generate the output at the moment) but I don't know how many users will take this on-the-fly approach instead of having prepared tiles. I may just not have enough context.

Since this code implicitly writes tiles with `tempfile` abstracted away from the user, an alternate temp dir could be a good argument to offer, allowing control over where tiles go as well as making it more apparent to an unsuspecting user that there is a potentially really large side effect to using this particular leaflet `add*` function. Possibly an `overwrite = FALSE` argument as well so that if a tiles directory by the same name already exists, recalling the same map display with raster layer in an exploratory, interactive R session could potentially skip the tile creation step on subsequent redraws of a map involving that tile set. This could help with the issue in the previous paragraph especially if someone is looking at multiple raster layers, but on the other hand, it's also just another way of moving closer to having the user just create multiple, named tile sets first and then draw them in leaflet as often as they want without having to repeatedly create them. It seems like the more seamless we make it, the more it begins to reflect a separation of the two processes. Thoughts?

**Comment by tim-salabim** [comment]
> In RStudio no map appears (in browser view), just the gray background with the zoom buttons in the corner. In the standard R GUI, simply nothing happens after addTiledRasterImage.

The map is set-up with `L.CRS.Simple` which is currently the only properly working crs for this. Also the raster doesn't have any projection information. I am a bit surprised that nothing happens in the standard R GUI as it should open a browser once the map call is triggered.

> It definitely a useful function (I'm just not able to generate the output at the moment) but I don't know how many users will take this on-the-fly approach instead of having prepared tiles. I may just not have enough context.

I agree that the time to create the tiles is in most cases where it makes sense too long! Therefore, I also doubt the usefulness of on-the-fly tile generation and visualisation. A far better approach would be to have on-the-fly tiling on the js side. But that is beyond my ability and willingness.

> Since this code implicitly writes tiles with tempfile abstracted away from the user, an alternate temp dir could be a good argument to offer, allowing control over where tiles go as well as making it more apparent to an unsuspecting user that there is a potentially really large side effect to using this particular leaflet add* function. Possibly an overwrite = FALSE argument as well so that if a tiles directory by the same name already exists, recalling the same map display with raster layer in an exploratory, interactive R session could potentially skip the tile creation step on subsequent redraws of a map involving that tile set.

That would, I guess, be possible but we would need to copy the folders to the tempfolder where the map is saved as leaflet only works with relative paths from the `index.html` file. 

In summary, I think I'm just gonna leave the fork where it is for the moment. I agree that all the effort is hardly worth it given that we need to wait for a long time to genereate meaningful tile sets (i.e. at high zoom levels). Especially, as we already have `plainView` in mapview which renders large rasters with acceptable speed.

**Comment by leonawicz** [comment]
This all sounds good. When I have time I will focus on the general issue with alternate projections, but even there I don't have any solutions yet, and no sign that I will have any time to devote to it soon. Luckily most people just want standard Mercator or I'd probably have received a lot more requests by now.

**Comment by leonawicz** [comment]
Closing since this issue is very old
